lib/ClangImporter/ClangImporter.cpp

```diff
@@ -9204,6 +9188,88 @@ swift::importer::getCxxRefConventionWithAttrs(const clang::Decl *decl) {
   return std::nullopt;
 }
 
+static const std::vector<std::vector<std::string>> ConflictingSwiftAttrs = {
+    {"Escapable", "~Escapable", "escapable_if:"},
+    {"~Copyable", "copyable_if:"},
+    {"mutating", "nonmutating"}};
+
+static inline bool isConditionalAttr(StringRef attrName) {
+  return attrName == "escapable_if:" || attrName == "copyable_if:";
+}
+
+void swift::importer::validateSwiftAttributes(
+    const clang::NamedDecl *decl, ClangImporter::Implementation &impl) {
+  auto transformToString = [](std::vector<std::string> &attrs) {
+    if (attrs.size() == 1)
+      return "'" + attrs[0] + "'";
+    std::string result;
+    llvm::raw_string_ostream os(result);
+    unsigned numAttrs = attrs.size();
+    for (unsigned i = 0; i < numAttrs; ++i) {
+      os << "'" << attrs[i] << "'";
+      if (i < numAttrs - 2)
+        os << ", ";
+      else if (i == numAttrs - 2)
+        os << " or ";
+    }
+    return result;
+  };
+
+  if (!decl->hasAttrs())
+    return;
+
+  for (auto groupOfAttrs : ConflictingSwiftAttrs) {
```

Comment:
Nit: I think we want to have `const auto&` here to avoid copying the smaller vectors and allocating.